Fix migration completion logic (along with other bugs)

acardace opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Before this PR:

KubeVirt would mark a migration as completed as soon as the domain would start running on the target, this is wrong as it doesn't really mean that the migration ended and it is particularly harmful when the migration runs in PostCopy mode as it would lead to hotplug operations to fail.

After this PR:

The migration code marks the migration as completed when it is actually completed, in addition this PR fixes a series of bugs in the migration logic.

A nice side-effect is that now hotplug works in postcopy-mode.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Fix virt-handler migration completion logic 
```

acardace commented:
This PR doesn't introduce any new unit tests, those are in the works, I first want to see how the e2e tests go.

jean-edouard commented on pkg/virt-handler/vm.go:
```diff
@@ -3447,6 +3405,8 @@ func (c *VirtualMachineController) finalizeMigration(vmi *v1.VirtualMachineInsta
 		return fmt.Errorf("%s: %v", errorMessage, err)
 	}
 
+	vmi.Status.MigrationState.Completed = true
```
I'm confused... In a previous commit, you say:
```
virt-launcher: remove Completed from MigrationMetadata
This field is no longer useful [...]
```
Why do we still use it here then?

vladikr replied:
Leftover perhaps?

acardace replied:
I removed it from the migrationMetadata, I'm not removing it from the VMI Status.

This is needed to ensure we don't call finalizeMigration in an endless loop.

jean-edouard reviewed (commented):
Amazing, thank you @acardace !
This fixes multiple things we've struggled with before.
Couple questions in-line, nothing major.
Please see this related PR: https://github.com/kubevirt/kubevirt/pull/13224 (your PR makes mine less critical, but I believe it still makes sense)

vladikr commented:
@acardace we need a bug to backport this work. Do we have any?

vladikr commented:
@awels fyi

vladikr commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -222,6 +223,7 @@ func newLibvirtDomainManager(connection cli.Connection, virtShareDir, ephemeralD
 		cancelSafetyUnfreezeChan: make(chan struct{}),
 		migrateInfoStats:         &stats.DomainJobInfo{},
 		metadataCache:            metadataCache,
+		setTimeOnce:              sync.Once{},
```
:+1:

vladikr commented:
> Amazing, thank you @acardace ! This fixes multiple things we've struggled with before. Couple questions in-line, nothing major. Please see this related PR: #13224 (your PR makes mine less critical, but I believe it still makes sense)

Wrt #13224 , I think we can merge the two (sorry @jean-edouard I completely lost track of it)
I like the errChan in 13224 but we also need to set the migration result as fast as possible - the moment migrateToURI3 exists without an error, as in this PR. (and maybe close the monitoring thread)

acardace commented on pkg/virt-launcher/virtwrap/live-migration-source.go:
```diff
@@ -374,13 +375,18 @@ func (l *LibvirtDomainManager) cancelMigration(vmi *v1.VirtualMachineInstance) e
+	if migrationMetadata.EndTimestamp != nil {
+		// the migration result has already been reported and should not be overwritten
+		return nil
+	}
```
Yes, you're right, I have to special case the abortion since it's a 2 step operation.

I'll delay putting the EndTimestamp once the abortion is completed.

acardace commented:
> > Amazing, thank you @acardace ! This fixes multiple things we've struggled with before. Couple questions in-line, nothing major. Please see this related PR: #13224 (your PR makes mine less critical, but I believe it still makes sense)
> 
> Wrt #13224 , I think we can merge the two (sorry @jean-edouard I completely lost track of it) I like the errChan in 13224 but we also need to set the migration result as fast as possible - the moment migrateToURI3 exists without an error, as in this PR. (and maybe close the monitoring thread)

I agree about closing the monitoring thread, @jean-edouard can you add this in your PR? I think it's a better fit there.
I think we can always send the return value of `migrateHelper` (even if it's nil) and then close the monitor as soon as we receive anything on the channel.

jean-edouard commented:
> > > Amazing, thank you @acardace ! This fixes multiple things we've struggled with before. Couple questions in-line, nothing major. Please see this related PR: #13224 (your PR makes mine less critical, but I believe it still makes sense)
> > 
> > 
> > Wrt #13224 , I think we can merge the two (sorry @jean-edouard I completely lost track of it) I like the errChan in 13224 but we also need to set the migration result as fast as possible - the moment migrateToURI3 exists without an error, as in this PR. (and maybe close the monitoring thread)
> 
> I agree about closing the monitoring thread, @jean-edouard can you add this in your PR? I think it's a better fit there. I think we can always send the return value of `migrateHelper` (even if it's nil) and then close the monitor as soon as we receive anything on the channel.

What do you mean? Take things from your PR to put in mine? Not sure I get the point. Feel free to do the opposite though!

acardace commented:
> > > > > Amazing, thank you @acardace ! This fixes multiple things we've struggled with before. Couple questions in-line, nothing major. Please see this related PR: #13224 (your PR makes mine less critical, but I believe it still makes sense)
> > > > 
> > > > 
> > > > Wrt #13224 , I think we can merge the two (sorry @jean-edouard I completely lost track of it) I like the errChan in 13224 but we also need to set the migration result as fast as possible - the moment migrateToURI3 exists without an error, as in this PR. (and maybe close the monitoring thread)
> > > 
> > > 
> > > I agree about closing the monitoring thread, @jean-edouard can you add this in your PR? I think it's a better fit there. I think we can always send the return value of `migrateHelper` (even if it's nil) and then close the monitor as soon as we receive anything on the channel.
> > 
> > 
> > What do you mean? Take things from your PR to put in mine? Not sure I get the point. Feel free to do the opposite though!
> 
> I meant that since you're changing things with the error channel inside the monitor you could put it in there to hopefully have less conflicts, but I can do it in mine as well.
> 
> We'll get merge conflicts anyway I guess.

@vladikr @jean-edouard I think I'll do this after Jed's PR gets merged, otherwise I'd have to copy his commits in this PR.

vladikr commented:
> > > > > > Amazing, thank you @acardace ! This fixes multiple things we've struggled with before. Couple questions in-line, nothing major. Please see this related PR: #13224 (your PR makes mine less critical, but I believe it still makes sense)
> > > > > 
> > > > > 
> > > > > Wrt #13224 , I think we can merge the two (sorry @jean-edouard I completely lost track of it) I like the errChan in 13224 but we also need to set the migration result as fast as possible - the moment migrateToURI3 exists without an error, as in this PR. (and maybe close the monitoring thread)
> > > > 
> > > > 
> > > > I agree about closing the monitoring thread, @jean-edouard can you add this in your PR? I think it's a better fit there. I think we can always send the return value of `migrateHelper` (even if it's nil) and then close the monitor as soon as we receive anything on the channel.
> > > 
> > > 
> > > What do you mean? Take things from your PR to put in mine? Not sure I get the point. Feel free to do the opposite though!
> > 
> > 
> > I meant that since you're changing things with the error channel inside the monitor you could put it in there to hopefully have less conflicts, but I can do it in mine as well.
> > We'll get merge conflicts anyway I guess.
> 
> @vladikr @jean-edouard I think I'll do this after Jed's PR gets merged, otherwise I'd have to copy his commits in this PR.

Yes, I agree. Let's get @jean-edouard 's PR in and then follow up here.

acardace commented:
/test pull-kubevirt-e2e-k8s-1.31-sig-compute-migrations

alicefr commented on pkg/virt-handler/vm.go:
```diff
@@ -628,7 +628,6 @@ func (c *VirtualMachineController) setMigrationProgressStatus(vmi *v1.VirtualMac
 		vmi.Status.MigrationState.EndTimestamp = migrationMetadata.EndTimestamp
 	}
 	vmi.Status.MigrationState.AbortStatus = v1.MigrationAbortStatus(migrationMetadata.AbortStatus)
-	vmi.Status.MigrationState.Completed = migrationMetadata.Completed
```
Is the the `Completed` field deprecated and not used anymore?

awels commented on pkg/virt-launcher/virtwrap/live-migration-source.go:
```diff
@@ -374,20 +375,30 @@ func (l *LibvirtDomainManager) cancelMigration(vmi *v1.VirtualMachineInstance) e
 		}
 	}
 
+	if migrationMetadata.EndTimestamp != nil {
```
Am I missing something, this looks identical to the code on line 385 and nothing has modified the migrationMetadata so this seems redundant.

acardace replied:
Yes, rebase mistake, thank you!

xpivarc commented on pkg/virt-launcher/virtwrap/live-migration-source.go:
```diff
@@ -399,8 +400,10 @@ func (l *LibvirtDomainManager) setMigrationResultHelper(failed bool, completed b
-			migrationMetadata.EndTimestamp = &now
+		}
+
+		if string(abortStatus) != string(v1.MigrationAbortInProgress) {
+			migrationMetadata.EndTimestamp = pointer.P(metav1.Now())
```
Abortion can also fail. In what case we need end timestamp and not completed?

acardace replied:
This is an abortion in progress though which is not the end of a migration, rather we're waiting for the abortion request to be processed so in this case I think we should not be setting the end timestamp.

xpivarc replied:
What I mean here is that you set the `EndTimestamp` even if the abortion fails. This seems wrong. Maybe we can add unit test?

acardace replied:
I think we should, the migration process has ended, even if it was aborted and the abortion failed. No?

xpivarc replied:
It might not be aborted and you will see abortion failed.

acardace replied:
Oh I see what you mean, `setMigrationAbortStatus` currently sets Failed to false while I though it'd set it to True, then yes it makes more sense to not set the EndTimestamp when the migration gets aborted.

acardace replied:
Just re-pushed, now abortion never sets the endTimestamp.

xpivarc commented on pkg/virt-launcher/virtwrap/live-migration-source.go:
```diff
@@ -360,7 +360,7 @@ func (l *LibvirtDomainManager) initializeMigrationMetadata(vmi *v1.VirtualMachin
 func (l *LibvirtDomainManager) cancelMigration(vmi *v1.VirtualMachineInstance) error {
 	migration, _ := l.metadataCache.Migration.Load()
-	if migration.Completed || migration.Failed || migration.StartTimestamp == nil {
+	if migration.Failed || migration.StartTimestamp == nil {
```
Don't you want to check for end timestamp?

acardace replied:
yes you're right.

xpivarc commented on pkg/virt-handler/vm_test.go:
```diff
@@ -1914,7 +1914,7 @@ var _ = Describe("VirtualMachineInstance", func() {
 				StartTimestamp: &now,
 				EndTimestamp:   &now,
-				Completed:      true,
+				Failed:         false,
```
This seems redundant

xpivarc replied:
@acardace Unresolved as I still see this in the commit.

acardace replied:
oops, let me fix it.

acardace replied:
Now it should be ok.

xpivarc commented on pkg/virt-launcher/virtwrap/live-migration-source.go:
```diff
@@ -375,7 +375,7 @@ func (l *LibvirtDomainManager) cancelMigration(vmi *v1.VirtualMachineInstance) e
 	return nil
 }
 
-func (l *LibvirtDomainManager) setMigrationResultHelper(failed bool, completed bool, reason string, abortStatus v1.MigrationAbortStatus) error {
```
I believe you still need the argument here to set the timestamp

acardace replied:
Why? That's implicit, if you're calling this function you're setting a migration result, so it means the migration has ended.

The abortion in progress is a special case which should probably be separated from this function.

xpivarc replied:
>The abortion in progress is a special case which should probably be separated from this function.

But as of now it is not so we should not set `EndTimestamp` on every call.

acardace replied:
It doesn't get set if the abortion is in progress.

xpivarc replied:
We have call `l.setMigrationAbortStatus(v1.MigrationAbortFailed)` as well and I think this is wrong or not?

xpivarc replied:
this is connected to https://github.com/kubevirt/kubevirt/pull/13729#discussion_r1926873812

xpivarc reviewed (commented):
Partial review so far

acardace commented on pkg/virt-handler/vm.go:
```diff
@@ -1741,6 +1744,16 @@ func (c *VirtualMachineController) getVMIFromCache(key string) (vmi *v1.VirtualM
+	// Get it from the global store as during a migration
+	// the VMI could disappear momentarily from both the source store
+	// and the target store
+	if !exists {
```
well if you get it from the target store then the VMI is not supposed to run on this node, so I'd say this assumption was valid even before this PR no? So we should be already protected from this.

vladikr replied:
@xpivarc I second that. if we get the vmi from the target it means it shouldn't run on this node.
I think the risk didn't change. we also check the migration UUID later on in the flow.

xpivarc replied:
The target store contains VMI only if is is supposed to run on this node as we have the label selector. Am I missing anything? 

>I think the risk didn't change. we also check the migration UUID later on in the flow.

There is slight difference here as previously you would only get the VMI if it is supposed to run on the node and now you get it even if it is supposed to run on another. Isn't it? But I admit I didn't check if the later logic is accommodating this

xpivarc replied:
Maybe we can add a unit test in this commit that would showcase what I described.

vladikr replied:
I wonder what's the benefit of checking three informers instead of just the `vmiStore` in this function.

Do we want to explicitly verify in this function that the VMI doesn't exist on this node or wasn't deleted in general?

@acardace could you remind me what was the original issue we needed this change for?

Looking at this code I'm not sure why the `vmiTargetStore` doesn't get eventually updated, so on the next iteration, we would do the right thing.

vladikr replied:
> The target store contains VMI only if is is supposed to run on this node as we have the label selector. Am I missing anything?
> 
> > I think the risk didn't change. we also check the migration UUID later on in the flow.
> 
> There is slight difference here as previously you would only get the VMI if it is supposed to run on the node and now you get it even if it is supposed to run on another. Isn't it? But I admit I didn't check if the later logic is accommodating this

@xpivarc hmm, this is not what was or is happening in this function as it was looking in both stores.

xpivarc replied:
>@acardace could you remind me what was the original issue we needed this change for?

The problem which was observed was that when we switch the VMI (change of node name) it is dropped from source cache but it might not arrive to the target domain yet.

xpivarc replied:
>@xpivarc hmm, this is not what was or is happening in this function as it was looking in both stores.

Can you elaborate? The store do use label selectors (well the informer's list do...)

vladikr replied:
> > @xpivarc hmm, this is not what was or is happening in this function as it was looking in both stores.
> 
> Can you elaborate? The store do use label selectors (well the informer's list do...)

Yes, it was looking in both stores, hoping to get any vmi object.

vladikr replied:
@acardace @xpivarc why do we still need to look here at all the stores (source, target, and the general vmi store)? Can't we only look at the general vmiStore?
If we want to know that the vmi object is still present (as has not been deleted), then looking at the vmiStore is enough. But if we explicitly want to know that it either on the target or the source, then this logic here is wrong and we should allow the vmi to appear on the target before we take any action on the source. 
It seems to me that this is enough to check the vmiStore at this point.

vladikr replied:
@xpivarc which filtering?
The current code seems to be redundant:

```
// Fetch the latest Vm state from cache
	obj, exists, err := c.vmiSourceStore.GetByKey(key)
	if err != nil {
		return nil, false, err
	}
	if !exists {
		obj, exists, err = c.vmiTargetStore.GetByKey(key)
		if err != nil {
			return nil, false, err
		}
	}

	// Get it from the global store as during a migration
	// the VMI could disappear momentarily from both the source store
	// and the target store
	if !exists {
		obj, exists, err = c.vmiStore.GetByKey(key)
		if err != nil {
			return nil, false, err
		}
	}
```

If the VMI is not in any of the stores it will look at the `vmiStore`, so what is the point of checking the vmiSource and the vmiTarget if we can simply look at the vmiStore here?!

Also, keep in mind that the VMI objects are being put into this queue as a result of an event that originated in the vmiSource, vmiTarget, and the domain informer so the objects are already "filtered"

xpivarc commented on pkg/virt-handler/vm.go:
```diff
@@ -509,42 +509,6 @@ func (c *VirtualMachineController) hasGracePeriodExpired(dom *api.Domain) (hasEx
 	return
 }
 
-func (c *VirtualMachineController) hasTargetDetectedReadyDomain(vmi *v1.VirtualMachineInstance) (bool, int64) {
```
FYI @vladikr @davidvossel 
Would you recollect why did we do this?

Thought: We could push this to separate PR

vladikr replied:
Isn't the explanation below satisfying?
I think it is to prevent a similar issue to what we see now in `getVMIFromCache` where the vmiTargetStore` is not aware of the VMI but the domain already pushed an event via the notifier.

xpivarc replied:
What isn't clear to me here is why we move the VMI to failed state as this could be just long informer lag...

davidvossel replied:
>What isn't clear to me here is why we move the VMI to failed state as this could be just long informer lag...

@xpivarc from the code

```
        // If we can't detect where the migration went to, then we have no
        // way of transferring ownership. The only option here is to move the
        // vmi to failed.  The cluster vmi controller will then tear down the
        // resulting pods.
```

I'm pretty sure i wrote that. My thought process was, if we ever got into a scenario where libvirt detected the domain on the target node, but that the source virt-handler never officially handed of the VMI to the target node, then the VMI is kind of in an unknow state. so i thought the logical thing was to move the VMI to a failed state since it's unclear how to recover from this scenario.


now, that was written a really long time ago. My understanding of the migration logic in libvirt may have been incomplete and you all may have thoughts on how we can recover this scenario more gracefully.

vladikr commented:
@acardace shouldn't `vmiStore` be part of the hasSynced function on line 226 in vm.go?

acardace commented:
> @acardace shouldn't `vmiStore` be part of the hasSynced function on line 226 in vm.go?

yes! good catch!

iholder101 commented on pkg/virt-launcher/virtwrap/live-migration-source.go:
```diff
@@ -389,18 +390,19 @@ func (l *LibvirtDomainManager) setMigrationResultHelper(failed bool, completed b
+		if abortStatus != "" {
+			migrationMetadata.AbortStatus = string(abortStatus)
+		} else {
+			// only mark the migration as complete if there was no abortion
```
why is that? shouldn't we (as the commit title suggests) always set an end timestamp?

acardace replied:
Abortion is a special case, I think it makes a bit more sense to not set it because with this PR setting the EndTimestamp basically means that the migration completed while with abortion this is never true.

Also read this https://github.com/kubevirt/kubevirt/pull/13729#discussion_r1923599852.

iholder101 commented on pkg/virt-launcher/virtwrap/manager.go:
```diff
@@ -345,16 +347,14 @@ func (l *LibvirtDomainManager) setGuestTime(vmi *v1.VirtualMachineInstance) erro
-	}
-
-	go func() {
+	go l.setTimeOnce.Do(func() {
```
I think the recommended way of doing something like that is:
```go
	l.setTimeOnce.Do(func() {
		go func() {
			// Actual logic
		}()
	})
```

The difference is that with the current code IIUC a new go routine will always be triggered, but will serve as a no-op after the first execution, while with my suggestion only a single go routine would ever be triggered.

acardace replied:
Nice, thank you for the suggestion, def better!

iholder101 commented on pkg/virt-handler/vm.go:
```diff
@@ -694,7 +694,6 @@ func (c *VirtualMachineController) migrationSourceUpdateVMIStatus(origVMI *v1.Vi
 		vmi.Status.NodeName = migrationHost
 		// clean the evacuation node name since have already migrated to a new node
 		vmi.Status.EvacuationNodeName = ""
-		vmi.Status.MigrationState.Completed = true
```
Can you also remove the comment above, which is now misleading?

```
		// this is the migration ACK.
		// At this point we know that the migration has completed and that
		// the target node has seen the domain event.
```

acardace replied:
The comment is still relevant, when that bit of code is executed the migration has actually ended.

iholder101 commented on pkg/virt-handler/vm.go:
```diff
@@ -1907,9 +1907,7 @@ func (c *VirtualMachineController) defaultExecute(key string,
 			log.Log.Object(vmi).V(3).Info("Deleting domain for VirtualMachineInstance with deletion timestamp.")
 			shouldDelete = true
 		default:
-			if vmi.IsFinal() {
-				shouldCleanUp = true
-			}
```
I think it makes sense to always cleanup if the vmi is deleted, however curious regarding to how this could happen. Why would a deleted VMI not be final?

acardace replied:
A vmi can be in scheduling/scheduled as well.

iholder101 reviewed (commented):
Thank you @acardace and kudos for finding these crucial issues! Well done!
Left some small comments below

acardace commented:
https://github.com/kubevirt/kubevirt/pull/13729#discussion_r1931810310 @iholder101 , in that commit I'm talking about the migrationMetadata (which is completely internal), the Status API field is still there.

xpivarc commented on pkg/virt-launcher/virtwrap/live-migration-source.go:
```diff
@@ -382,11 +382,15 @@ func (l *LibvirtDomainManager) setMigrationResultHelper(failed bool, reason stri
 		if metaAbortStatus == string(abortStatus) && metaAbortStatus == string(v1.MigrationAbortInProgress) {
 			return domainerrors.MigrationAbortInProgressError
 		}
+	} else if metaAbortStatus == string(v1.MigrationAbortInProgress) &&
```
This branch will be taken only if `abortStatus` is empty or am I missing something?

acardace replied:
Fixed this, I made it simpler.

xpivarc commented:
Overall I think this PR is converging but I would feel more comfortable to split this into more pieces. Primarily for easier reviews and backports

acardace commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -240,6 +240,7 @@ func (app *virtHandlerApp) Run() {
 	// Wire VirtualMachineInstance controller
 	factory := controller.NewKubeInformerFactory(app.virtCli.RestClient(), app.virtCli, nil, app.namespace)
 
+	vmiInformer := factory.VMI()
```
Done, thank you, will push it shortly.

acardace commented on pkg/virt-handler/vm.go:
```diff
@@ -652,6 +624,27 @@ func (c *VirtualMachineController) migrationSourceUpdateVMIStatus(origVMI *v1.Vi
 	}
+}
+
+func (c *VirtualMachineController) migrationSourceUpdateVMIStatus(origVMI *v1.VirtualMachineInstance, domain *api.Domain) error {
```
Will do.

acardace commented on pkg/virt-handler/vm_test.go:
```diff
@@ -1925,55 +1924,59 @@ var _ = Describe("VirtualMachineInstance", func() {
 			client.EXPECT().Ping().AnyTimes()
-			client.EXPECT().FinalizeVirtualMachineMigration(gomock.Any(), gomock.Any())
+			client.EXPECT().FinalizeVirtualMachineMigration(gomock.Any(), gomock.Any()).Return(nil)
+			mockHotplugVolumeMounter.EXPECT().Mount(gomock.Any(), mockCgroupManager).Return(nil)
+			mockHotplugVolumeMounter.EXPECT().Unmount(gomock.Any(), mockCgroupManager).Return(nil)
```
We already cleared this up offline, tldr: this unit test exercise the target node finalization the migration.

acardace commented on pkg/virt-handler/vm_test.go:
```diff
@@ -1925,55 +1924,59 @@ var _ = Describe("VirtualMachineInstance", func() {
+			client.EXPECT().FinalizeVirtualMachineMigration(gomock.Any(), gomock.Any()).Return(nil)
+			mockHotplugVolumeMounter.EXPECT().Mount(gomock.Any(), mockCgroupManager).Return(nil)
+			mockHotplugVolumeMounter.EXPECT().Unmount(gomock.Any(), mockCgroupManager).Return(nil)
+			client.EXPECT().SyncVirtualMachine(gomock.Any(), gomock.Any())
```
We already cleared this up offline, tldr: this unit test exercise the target node finalizing the migration.

acardace commented on pkg/virt-handler/vm_test.go:
```diff
@@ -2021,13 +2027,14 @@ var _ = Describe("VirtualMachineInstance", func() {
 			client.EXPECT().Ping().AnyTimes()
 			client.EXPECT().FinalizeVirtualMachineMigration(gomock.Any(), gomock.Any())
 			client.EXPECT().SyncVirtualMachineCPUs(gomock.Any(), gomock.Any())
+			mockHotplugVolumeMounter.EXPECT().Mount(gomock.Any(), mockCgroupManager).Return(nil)
+			mockHotplugVolumeMounter.EXPECT().Unmount(gomock.Any(), mockCgroupManager).Return(nil)
```
We already cleared this up offline, tldr: this unit test exercise the target node finalizing the migration.

acardace commented on tests/migration/migration.go:
```diff
@@ -1560,9 +1560,6 @@ var _ = SIGMigrationDescribe("VM Live Migration", decorators.RequiresTwoSchedula
 				vmi = libmigration.ConfirmVMIPostMigrationFailed(vmi, migrationUID)
-				// Not sure how consistent the entire error is, so just making sure the failure happened in libvirt. Example string:
-				// Live migration failed error encountered during MigrateToURI3 libvirt api call: virError(Code=1, Domain=7, Message='internal error: client socket is closed'
-				Expect(vmi.Status.MigrationState.FailureReason).To(ContainSubstring("libvirt api call"))
```
Like the comment says, the error is not reliable.

acardace commented on pkg/virt-launcher/virtwrap/live-migration-source.go:
```diff
@@ -989,9 +982,14 @@ func (l *LibvirtDomainManager) migrateHelper(vmi *v1.VirtualMachineInstance, opt
 		return fmt.Errorf("error encountered during MigrateToURI3 libvirt api call: %v", err)
 	}
 
+	log.Log.Object(vmi).Errorf("migration completed successfully")
```
I'd strongly keep it, this is part of the fix otherwise the migration monitor might change the migration result unexpectedly and cause failures in e2e/unit tests.

acardace commented:
@xpivarc please take a second look, I've removed 2 of the commits you asked for (they will be in a separate PR), and I fixed the upgrade tests by making the target node set the migration endTimestamp.

@vladikr please take look as well.

iholder101 commented:
> Overall I think this PR is converging but I would feel more comfortable to split this into more pieces. Primarily for easier reviews and backports

+1
I think the PR in its current form is hard to review, and will probably be hard to backport.
I also think it would be beneficial to split it into multiple PRs.

jean-edouard commented:
/close
Replaced by https://github.com/kubevirt/kubevirt/pull/14705
